File: Advanced/PageAndRDL/CustomPdfExport/VB.NET/PdfRendering/Rendering/PdfGenerator.vb
Imports PdfSharp.Pdf
Imports PdfSharp.Drawing
Imports GrapeCity.ActiveReports.Rendering.Export
Imports GrapeCity.ActiveReports.Rendering
Imports GrapeCity.ActiveReports.Drawing
Imports GrapeCity.ActiveReports.Extensibility.Rendering.Components
Imports GrapeCity.ActiveReports.Extensibility.Rendering
Imports System.IO
Imports System.Drawing
Imports GrapeCity.ActiveReports.Drawing.Gdi

''' <summary>
''' PDF generator based on PDFsharp: http://www.pdfsharp.net/
''' </summary>
Friend NotInheritable Class PdfGenerator
	Implements IGenerator
	Implements IDisposable
	Private Shared ReadOnly GdiMetricsProvider As ITextMetricsProvider = New TextMetricsProvider()
	Private Shared ReadOnly GdiRenderersFactory As IRenderersFactory = New GdiRenderersFactory()

	Private ReadOnly _outputStream As Stream

	Private _document As PdfDocument
	Private ReadOnly _graphics As IList(Of XGraphics) = New List(Of XGraphics)()
	Private ReadOnly _images As PdfImagesFactory
	Private ReadOnly _fonts As PdfFontsFactory
	Private ReadOnly _outlines As IDictionary(Of String, PdfOutline) = New Dictionary(Of String, PdfOutline)()

	Public Sub New(outputStream As Stream, embedFonts As Boolean)
		_outputStream = outputStream

		_images = New PdfImagesFactory()
		_fonts = New PdfFontsFactory(embedFonts)
	End Sub

	Sub Dispose() Implements IDisposable.Dispose
		For Each g As XGraphics In _graphics
			Debug.Assert(g.GraphicsStateLevel = 0)
			g.Dispose()
		Next
		_graphics.Clear()
		If _document IsNot Nothing Then
			_document.Save(_outputStream)
			_document.Dispose()
		End If
		_document = Nothing
		CType(_images, IDisposable).Dispose()
	End Sub

	ReadOnly Property IsDelayedContentSupported() As Boolean Implements IGenerator.IsDelayedContentSupported
		Get
			Return True
		End Get
	End Property

	Public ReadOnly Property ProfessionalEdition() As Boolean Implements IGenerator.ProfessionalEdition
		Get
			Return False
		End Get
	End Property

	Public ReadOnly Property InteractivitySupport() As InteractivityType Implements IGenerator.InteractivitySupport
		Get
			Return InteractivityType.BuiltinHyperlinks
		End Get
	End Property

	Public ReadOnly Property MetricsProvider() As ITextMetricsProvider Implements IGenerator.MetricsProvider

		Get
			Return GdiMetricsProvider
		End Get
	End Property

	Public ReadOnly Property RenderersFactory() As IRenderersFactory Implements IGenerator.RenderersFactory
		Get
			Return GdiRenderersFactory
		End Get
	End Property

	Sub Init(report As IReport) Implements IGenerator.Init
		_document = New PdfDocument() With {
				.Version = 17
			}
	End Sub

	Function NewPage(pageNumber As Integer, pageSize As SizeF) As IDrawingCanvas Implements IGenerator.NewPage
		Dim page As New PdfPage(_document)
		page.Width = New XUnit(pageSize.Width / PdfConverter.TwipsPerPoint, XGraphicsUnit.Point)
		page.Height = New XUnit(pageSize.Height / PdfConverter.TwipsPerPoint, XGraphicsUnit.Point)
		_document.AddPage(page)
		Dim graphics = XGraphics.FromPdfPage(page)
		_graphics.Add(graphics)
		Return New PdfContentGenerator(graphics, _images, _fonts)
	End Function

	Sub StartOutline(key As String, parentKey As String, targetPage As Integer, targetArea As RectangleF, name As String) Implements IGenerator.StartOutline
		Dim outline As New PdfOutline(name, _document.Pages(targetPage - 1)) With {
				.PageDestinationType = PdfPageDestinationType.Xyz,
				.Left = targetArea.X / PdfConverter.TwipsPerPoint,
				.Top = targetArea.Y / PdfConverter.TwipsPerPoint
			}
		Dim parentOutline As PdfOutline = Nothing
		If _outlines.TryGetValue(parentKey, parentOutline) Then
			parentOutline.Outlines.Add(outline)
		Else
			_document.Outlines.Add(outline)
		End If
		_outlines(key) = outline
	End Sub

	Sub AddBookmark(key As String, sourcePage As Integer, sourceArea As RectangleF, targetPage As Integer, targetArea As RectangleF) Implements IGenerator.AddBookmark
		Dim sourceRect = _graphics(sourcePage).Transformer.WorldToDefaultPage(sourceArea)
		_document.Pages(sourcePage).AddDocumentLink(New PdfRectangle(sourceRect), targetPage)
	End Sub

	Sub UrlGoTo(link As String, sourcePage As Integer, sourceArea As RectangleF) Implements IGenerator.UrlGoTo
		_document.Pages(sourcePage - 1).AddWebLink(New PdfRectangle(PdfConverter.Convert(sourceArea)), link)
	End Sub

	Sub AddSorting(key As String, sourcePage As Integer, sourceArea As RectangleF) Implements IGenerator.AddSorting
	End Sub

	Sub AddToggle(key As String, sourcePage As Integer, sourceArea As RectangleF) Implements IGenerator.AddToggle
	End Sub

	Sub BookmarkGoTo(key As String, sourcePage As Integer, sourceArea As RectangleF) Implements IGenerator.BookmarkGoTo
	End Sub

	Sub DrillthroughGoTo(reportName As String, parameters As IDictionary(Of String, Object), sourcePage As Integer, sourceArea As RectangleF) Implements IGenerator.DrillthroughGoTo
	End Sub

	Sub EndOutline(key As String) Implements IGenerator.EndOutline
	End Sub

	Sub FinishLinks() Implements IGenerator.FinishLinks
	End Sub
End Class

File: Advanced/PageAndRDL/CustomPdfExport/C#/PdfRendering/Rendering/PdfGenerator.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Drawing;
using System.IO;
using GrapeCity.ActiveReports.Extensibility.Rendering;
using GrapeCity.ActiveReports.Extensibility.Rendering.Components;
using GrapeCity.ActiveReports.Drawing;
using GrapeCity.ActiveReports.Rendering;
using GrapeCity.ActiveReports.Rendering.Export;
using PdfSharp.Drawing;
using PdfSharp.Pdf;
using GrapeCity.ActiveReports.Drawing.Gdi;

namespace GrapeCity.ActiveReports.Samples.Export.Rendering.PdfSharp
{
	/// <summary>
	/// PDF generator based on PDFsharp: http://www.pdfsharp.net/
	/// </summary>
	internal sealed class PdfGenerator : IGenerator, IDisposable
	{
		private static readonly ITextMetricsProvider GdiMetricsProvider = new TextMetricsProvider();
		private static readonly IRenderersFactory GdiRenderersFactory = new GdiRenderersFactory();

		private readonly Stream _outputStream;

		private PdfDocument _document;
		private readonly IList<XGraphics> _graphics = new List<XGraphics>();
		private readonly PdfImagesFactory _images;
		private readonly PdfFontsFactory _fonts;
		private readonly IDictionary<string, PdfOutline> _outlines = new Dictionary<string, PdfOutline>();

		public PdfGenerator(Stream outputStream, bool embedFonts)
		{
			_outputStream = outputStream;

			_images = new PdfImagesFactory();
			_fonts = new PdfFontsFactory(embedFonts);
		}

		void IDisposable.Dispose()
		{
			foreach (var g in _graphics)
			{
				Debug.Assert(g.GraphicsStateLevel == 0);
				g.Dispose();
			}
			_graphics.Clear();
			if (_document != null)
			{
				_document.Save(_outputStream);
				_document.Dispose();
			}
			_document = null;
			((IDisposable)_images).Dispose();
		}
		
		bool IGenerator.IsDelayedContentSupported
		{
			get { return true; }
		}

		public bool ProfessionalEdition
		{
			get { return false; }
		}

		public InteractivityType InteractivitySupport
		{
			get { return InteractivityType.BuiltinHyperlinks; }
		}

		public ITextMetricsProvider MetricsProvider
		{
			get { return GdiMetricsProvider; }
		}

		public IRenderersFactory RenderersFactory
		{
			get { return GdiRenderersFactory; }
		}

		void IGenerator.Init(IReport report)
		{
			_document = new PdfDocument { Version = 17 };
		}

		IDrawingCanvas IGenerator.NewPage(int pageNumber, SizeF pageSize)
		{
			var page = new PdfPage(_document);
			page.Width = new XUnit(pageSize.Width / PdfConverter.TwipsPerPoint, XGraphicsUnit.Point);
			page.Height = new XUnit(pageSize.Height / PdfConverter.TwipsPerPoint, XGraphicsUnit.Point);
			_document.AddPage(page);
			var graphics = XGraphics.FromPdfPage(page);
			_graphics.Add(graphics);
			return new PdfContentGenerator(graphics, _images, _fonts);
		}

		void IGenerator.StartOutline(string key, string parentKey, int targetPage, RectangleF targetArea, string name)
		{
			var outline = new PdfOutline(name, _document.Pages[targetPage - 1])
			{
				PageDestinationType = PdfPageDestinationType.Xyz,
				Left = targetArea.X / PdfConverter.TwipsPerPoint,
				Top = targetArea.Y / PdfConverter.TwipsPerPoint
			};
			PdfOutline parentOutline;
			if (_outlines.TryGetValue(parentKey, out parentOutline))
				parentOutline.Outlines.Add(outline);
			else
				_document.Outlines.Add(outline);
			_outlines[key] = outline;
		}
		
		void IGenerator.AddBookmark(string key, int sourcePage, RectangleF sourceArea, int targetPage, RectangleF targetArea)
		{
			var sourceRect = _graphics[sourcePage].Transformer.WorldToDefaultPage(XRect.FromLTRB(sourceArea.Left, sourceArea.Top, sourceArea.Right, sourceArea.Bottom));
			_document.Pages[sourcePage].AddDocumentLink(new PdfRectangle(sourceRect), targetPage);
		}
		
		void IGenerator.UrlGoTo(string link, int sourcePage, RectangleF sourceArea)
		{
			_document.Pages[sourcePage - 1].AddWebLink(new PdfRectangle(PdfConverter.Convert(sourceArea)), link);
		}
		
		void IGenerator.AddSorting(string key, int sourcePage, RectangleF sourceArea)
		{
		}

		void IGenerator.AddToggle(string key, int sourcePage, RectangleF sourceArea)
		{
		}

		void IGenerator.BookmarkGoTo(string key, int sourcePage, RectangleF sourceArea)
		{
		}

		void IGenerator.DrillthroughGoTo(string reportName, IDictionary<string, object> parameters, int sourcePage, RectangleF sourceArea)
		{
		}

		void IGenerator.EndOutline(string key)
		{
		}

		void IGenerator.FinishLinks()
		{
		}
	}
}
